ENH/BUG: Code Clean Up, Optimization, Documentation

TactfulDeity opened the pull request:
Request:
I would like to request a squash commit if this goes in, I ran the clang formatting with style `file`, but it doesn't pick up all of it that actions bot does hence the needless third commit.

Preface:
I only did the calc, core, and io folders completely, but I touched on other smaller plugins as well. I am not overly familiar with QT so I was cautious about removing those headers. I only did what I could finish over the long weekend, but I am willing to do more down the line. Special care should be taken reviewing `qtaim` plugin, I consolidated and cleaned up a lot, so I want to make sure I didn't break it.

Fixes:
#1776

List of Changes:
- Concatenated namespaces
- removed unneeded includes
- changed bool declarations and if checks to use lazy evaluation and cut down on needless branching
- changed places where ADL was not neccessary and added documentation and scoping to places that may benefit
- extensive changes to `qtaim` plugin (should be reviewed with scrutiny)
- removed `using std::namespace`
- changed `typedef` to `using` statements
- changed codes to remove obvious excessive copies
- fixed auto declarations to include pointer/ref
- extensive changes to default constructors/destructors
- documentation in the form of comments where appropriate
- small changes to prefer c++ standard structures where appropriate

Note:
I went through this in order to familiarize myself with the code base better, during which I noticed there are a bunch of "utility" classes that basically act as an interface of sorts. I think these could be changed to instead be namespaces to remove unneeded object overhead. Example: `atomutilities.h`. I will add these to the list of changes I can help with if other maintainers agree. 

Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

TactfulDeity commented:
I suspect it will need some include fixes for each of the platforms before going in, I will finish patching tomorrow and mark it ready for review after I get it passing.

ghutchis commented:
I can definitely do a squash commit, but I've also had a few that I should have squashed in the commit history.

I'll take a look tomorrow, thanks.

ghutchis commented:
BTW, GitHub is a little weird - you have to have "Fixes #1776" on one line for it to recognize the issue. I manually marked it though.

TactfulDeity commented:
Looks ready to go it passed ARM I will check back in and make any changes if another issue crops up in the bots.

ghutchis commented:
Is this ready for review? If not, `@` me when you think it's ready to examine. Thanks - definitely a big help.

TactfulDeity commented:
Not yet I am going to squash the last two bugs picked up by the address sanitizer and undefined behavior sanitizer as soon as I get out of work today

TactfulDeity commented:
I will definitely '@' you as soon as I get it working. Sorry its taking so long, I don't have a lot of time during the week.

ghutchis commented:
Oh, no worries. This isn't taking "long." And all of us are doing this in our "free" time. 😄 

Much appreciated whenever you get to it.

ghutchis commented on avogadro/core/color3f.h:
```diff
@@ -81,21 +82,15 @@ class Color3f
+inline Color3f::Color3f(float r, float g, float b) : m_data({ r, g, b }) {}
 
 inline Color3f::Color3f(int r, int g, int b)
+  : m_data({ static_cast<float>(r) / 255.0f, static_cast<float>(g) / 255.0f,
```
Did `clang` format this? It's one place I might lean to 3 separate lines for readability. 

(I'm also wondering if 255.0f should be a declared constant somewhere also for readability.)

TactfulDeity replied:
Yea I ran clang formatting over the commit with git and made all the changes posted by the formatting bot. 

I can absolutely extract it out into a constant and store it in the source file.

ghutchis commented on avogadro/core/crystaltools.cpp:
```diff
@@ -256,6 +256,9 @@ bool CrystalTools::niggliReduce(Molecule& molecule, Options opts)
 */
 #define NIGGLI_DEBUG(step)
 
+  // Allow ADL for swap
```
Might be worth expanding the acronym for scientists 😄

ghutchis commented on avogadro/core/gaussiansettools.cpp:
```diff
@@ -127,10 +121,7 @@ double GaussianSetTools::calculateSpinDensity(const Vector3& position) const
-    return true;
-  else
-    return false;
+  return m_molecule && dynamic_cast<GaussianSet*>(m_molecule->basisSet());
```
Let's put parens around this since it's essentially an assignment

ghutchis commented on avogadro/core/graph.cpp:
```diff
@@ -545,7 +539,7 @@ std::vector<std::set<size_t>> Graph::connectedComponents() const
   updateSubgraphs();
   std::vector<std::set<size_t>> r;
-  for (std::set<size_t> s : m_subgraphToVertices) {
+  for (const std::set<size_t>& s : m_subgraphToVertices) {
```
👍

ghutchis commented on avogadro/core/graph.cpp:
```diff
@@ -363,8 +357,8 @@ void Graph::removeEdges(size_t index)
   const std::vector<size_t>& edges = m_edgeMap[index];
-  for (size_t i = 0; i < edges.size(); ++i)
-    removeEdge(edges[i]);
+  for (size_t edge : edges)
```
👍

ghutchis commented on avogadro/core/graph.cpp:
```diff
@@ -562,7 +556,7 @@ size_t Graph::subgraphsCount() const
   updateSubgraphs();
   size_t r = 0;
-  for (std::set<size_t> s : m_subgraphToVertices) {
+  for (const std::set<size_t>& s : m_subgraphToVertices) {
```
👍

ghutchis commented:
Let me finish the review - a few other minor notes... but overall, it looks good.

ghutchis commented on avogadro/core/ringperceiver.cpp:
```diff
@@ -36,12 +37,20 @@
   memset(m_values, 0, size * size * sizeof(size_t));
 }
 
+DistanceMatrix::DistanceMatrix(const DistanceMatrix& other)
```
This code isn't used much, but this seems like a good idea (adding a copy assignment operator).

ghutchis commented on avogadro/core/slaterset.h:
```diff
@@ -141,6 +140,7 @@ class AVOGADROCORE_EXPORT SlaterSet : public BasisSet
   MatrixX& normalizedMatrix() { return m_normalized; }
   MatrixX& densityMatrix() { return m_density; }
 
+  // Does nothing?
```
In general, this type of basis set isn't used much. Like maybe 0.01% of the time. We've included it for backwards compatibility with a few programs.

ghutchis commented on avogadro/core/slatersettools.cpp:
```diff
@@ -74,35 +65,30 @@ double SlaterSetTools::calculateSpinDensity(const Vector3&) const
-    return true;
-  else
-    return false;
+  return m_molecule && dynamic_cast<SlaterSet*>(m_molecule->basisSet());
```
Let's add parens here.

ghutchis commented on avogadro/core/slatersettools.cpp:
```diff
@@ -74,35 +65,30 @@ double SlaterSetTools::calculateSpinDensity(const Vector3&) const
-    return true;
-  else
-    return false;
+  return val > -1e-20 && val < 1e-20;
```
Maybe better to just switch to `std::abs(val) < 1e-20`?

ghutchis commented on avogadro/qtplugins/apbs/apbs.cpp:
```diff
@@ -4,6 +4,7 @@
 ******************************************************************************/
 
 #include "apbs.h"
+
```
???

ghutchis replied:
Not sure why there's an extra blank line here.

TactfulDeity replied:
I added it to separate the header for the source from the local module includes following the recommended include ordering:

1. header with definitions
2. local includes
3. wider project includes
4. 3rd party libraries includes
5. standard library includes

This cuts down on include bloat because you include the most common ones last ensuring you can make the most of the already included libraries from the more niche files you would have to include either way.

TactfulDeity replied:
@ghutchis its just for readability and easy distinction

TactfulDeity commented:
@ghutchis I actually did have a question about the code. I am working on trying to fix whatever bug I introduced to graph.cpp.

In the Ubuntu address sanitizer run you can see it's actually failing a test case "connected components" (line 86 of Run tests).

However, I don't actually know if the result is wrong or the test case is wrong. I drew out a graph for the test case and it starts like this:

[0] [1] [2] [3] [4] [5]

Where each bracketed number is a node/vertex.

It ends (right before the assert after the final edge removal) looking like this:

[0]-[1]-[2] [3] [4]-[5]

Where each "-" represents an edge.

For reference the test case is on line 146 of graphtest.cpp.

The assert asks for the connected components count and says it should be "4". I am confused about where the 4 is coming from. I think it may be due to some part of how the implementation handles lone nodes or separate subgraphs? Any insight you may have would be greatly appreciated. I feel if I can understand the intent I can identify the bug better.

I poured over my changes to graph (and the member functions in general) a couple times yesterday but couldn't figure out what would cause this. I reverted the ADL on swap thinking maybe it was calling an unintended method that wasn't doing what was expected, but the issue persisted in the run from my most recent commit.

Thanks for all your time and consideration in helping me get spun up on the project.

TactfulDeity commented:
@ghutchis okay, I am out of work and actively working on making review changes. Did you see my message higher up about the graph bug?

ghutchis commented:
> @ghutchis I actually did have a question about the code. I am working on trying to fix whatever bug I introduced to graph.cpp.

Not sure I'll have a chance to look tonight, but I'll take a look tomorrow.

TactfulDeity commented:
Alright that concludes the review related patches, none of which touched the graph file, so the test is likely still failing for some reason beyond me. The sanitizer and undefined behavior were skipped for some reason so I am caching the error from previous commit here:

```
11/37 Test #11: Core-Graph .......................***Failed    0.05 sec
Running main() from /home/runner/work/avogadrolibs/build/thirdparty/gtest-prefix/src/gtest/googletest/src/gtest_main.cc
Note: Google Test filter = GraphTest.*
[==========] Running 11 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 11 tests from GraphTest
[ RUN      ] GraphTest.size
[       OK ] GraphTest.size (0 ms)
[ RUN      ] GraphTest.setSize
[       OK ] GraphTest.setSize (0 ms)
[ RUN      ] GraphTest.isEmpty
[       OK ] GraphTest.isEmpty (0 ms)
[ RUN      ] GraphTest.clear
[       OK ] GraphTest.clear (0 ms)
[ RUN      ] GraphTest.addVertex
[       OK ] GraphTest.addVertex (0 ms)
[ RUN      ] GraphTest.removeVertex
[       OK ] GraphTest.removeVertex (0 ms)
[ RUN      ] GraphTest.vertexCount
[       OK ] GraphTest.vertexCount (0 ms)
[ RUN      ] GraphTest.addEdge
[       OK ] GraphTest.addEdge (0 ms)
[ RUN      ] GraphTest.removeEdge
[       OK ] GraphTest.removeEdge (0 ms)
[ RUN      ] GraphTest.edgeCount
[       OK ] GraphTest.edgeCount (0 ms)
[ RUN      ] GraphTest.connectedComponents
/home/runner/work/avogadrolibs/avogadrolibs/avogadrolibs/tests/core/graphtest.cpp:169: Failure
Expected equality of these values:
  graph.connectedComponents().size()
    Which is: 5
  static_cast<size_t>(4)
    Which is: 4
[  FAILED  ] GraphTest.connectedComponents (0 ms)
[----------] 11 tests from GraphTest (1 ms total)

[----------] Global test environment tear-down
[==========] 11 tests from 1 test suite ran. (1 ms total)
[  PASSED  ] 10 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] GraphTest.connectedComponents

 1 FAILED TEST
```

ghutchis commented:
I had to have a moment to step through the test again. Here's what it should look like at the end of the test:
<img width="531" alt="Screenshot 2024-11-15 at 7 42 23 PM" src="https://github.com/user-attachments/assets/c707f01c-6ad5-492d-95d6-66897f298150">

(i.e., it removed all edges around node 4)

So yeah, it's a little weird that it's now reporting `5`.

I also don't see anything immediately obvious in what you've done. I'll look through more carefully.

ghutchis commented on avogadro/core/graph.cpp:
```diff
@@ -363,8 +357,8 @@ void Graph::removeEdges(size_t index)
     m_subgraphDirty[m_vertexToSubgraph[index]] = true;
 
   const std::vector<size_t>& edges = m_edgeMap[index];
-  for (size_t i = 0; i < edges.size(); ++i)
```
Here's the bug. If you put this back, the test will pass:
```
  for (size_t i = 0; i < edges.size(); ++i)
    removeEdge(edges[i]);
```

I always get nervous with iterators that remove things. I'd have to think carefully about why the range iterator is failing, but the simple solution is to just revert to the old-school version.

ghutchis commented:
Otherwise, the PR looks good. Fix the weird iterator and we're good to merge.

TactfulDeity commented:
@ghutchis Sweet, thanks so much for all your help. I'll get a fix in in a few hours before I head to bed tonight, so it should be ready to go (checks passed) in the AM.

TactfulDeity commented:
Alright fix is in and ready to go
